### ISSUE TITLE
Importing with union constructors in scope

### ISSUE THREAD

**Issue by Gabriel439** [dhall-lang/dhall-lang#994]
This is based on a discussion with @drewolson in Slack on alternative ways to solve the problem mentioned in https://github.com/dhall-lang/dhall-lang/issues/989 of ergonomically importing a union type.  I chose to create a new issue because I didn't want to derail the existing discussion on #989.

One idea @drewolson suggested is to add support for an `as UnionType` feature, like this:

```dhall
let Example = < prod | staging | development >

in  env:ENV as Example
```

... then the Dhall expression imported from the `ENV` environment variable would have the union constructors in scope.  So the environment variable would only need to be `ENV=prod`, for example.

A related idea, similar to one proposed by @Reisen a while back in https://github.com/dhall-lang/dhall-haskell/issues/676 is to bring local variables into scope (either wholesale or selectively) for the imported expression, like this:

```dhall
let prod = Example.prod
let staging = Example.staging
let development = Example.development
in  env:ENV context { prod, staging, development }  -- This would be the selective approach
```

The main issue I can foresee with both approaches is that the contents of the import would no longer type-check in isolation.  One possible variation on the above ideas is to only permit this feature for environment variables (which are more ergonomically constrained and less likely to be interpreted in isolation).

**Comment by sjakobi** [comment]
I like the `env:X as UnionType` idea!

> The main issue I can foresee with both approaches is that the contents of the import would no longer type-check in isolation.

Users could combine the import with an alternative, right? E.g.

```dhall
env:ENV as Example ? Example.prod
```

> One possible variation on the above ideas is to only permit this feature for environment variables (which are more ergonomically constrained and less likely to be interpreted in isolation).

What do you mean by "less likely to be interpreted in isolation", @Gabriel439?

I currently suspect that users might also find this feature useful for file or remote imports, although I haven't come up with a concrete example yet…

**Comment by Gabriel439** [comment]
@sjakobi: What I meant is that if one did something like this:

```dhall
let x = 1

in  ./example.dhall scope { x }
```

... and `./example.dhall` contained:

```dhall
x
```

... then `./example.dhall` would no longer type-check in isolation.

Or in other words, so far all Dhall imports have the property that they can be validated ahead-of-time for internal inconsistencies, but if they depend on variables that are only in scope at the import site then this is no longer the case.

The reason I suggested making an exception for environment variables is because it's unlikely that they would be validated ahead of time (since their use case is more "ephemeral")

**Comment by lisael** [comment]
I've recently got a case where I want an optional to be either `Some env:VAR as Text` or `None Text` if the environment variable is not set. It's not exactly the same use case, but I feel like it could be standarized in the same round.

My workaround is:
```hs
let var = env:VAR ? None Text
...
```
with this in my shell:
```sh
export VAR='Some "hello"'
```
Which is a little weird from a user standpoint.

**Comment by Nadrieril** [comment]
I think you can do:
```dhall
let var = Some env:VAR ? None Text
...
```
(you may need some more parentheses I can never remember precedences)

**Comment by Gabriel439** [comment]
I believe if you want to import the variable as `Text` and produce an `Optional` result, the exact syntax would be:

```dhall
Some (env:VAR as Text) ? None Text
```

**Comment by lisael** [comment]
oops. was easy. What tricks me is the mix of imports and otherwise expression. I read this as
```hs
Some (env:VAR as Text ? None Text)
```
In my mind `?` is an import operator and the lhs of `?` is the import before it. It's wrong, `?` only select the expression that resolves and type-checks correctly. Thanks.

**Comment by Gabriel439** [comment]
@lisael: Yeah, the `?` operator accepts arbitrary Dhall expressions on either side (although it's only really useful if the left-hand side has some import embedded within it).  So you could do something like `1 ? 2` if you wanted to, which would always evaluate to `1`.

**Comment by ymph** [comment]
Hello,

I really would like to see this feature implemented, even if limited to environment variables.

The workaround I found at the moment is to use construction like this:

```shell
export VAR="Val1"
export VAR_DEF="let EnumType : Type = < Val1 | Val2 | Val3 | Val4 > in EnumType.$VAR"

dhall <<< "let var=env:VAR_DEF in var"
```

But I find it very ugly. The type definition is located in another place than the code that really uses it, it makes the reuse of the enum type questionable and the environment is filled with esoteric types definitions.
All other workaround I can think of are a variant of the example and do not really answer the concerns.

But the last comment was made 3 years ago. Perhaps there are some more "standard" workaround used by the community.

Thanks !

**Comment by mattpolzin** [comment]
@Gabriella439 I know just a small amount about the strategy for accepting changes into the Dhall language specification (e.g. voting amongst those with Dhall implementations native to another programming language); I am curious whether this idea (and the alternative idea in #989) are kind of in limbo or if there's been some movement by you and the community towards one alternative or the other?

I can't promise anything, especially as a Haskell novice, but taking a stab at something like implementing the feature in this ticket would potentially be fun for me, though I'd be hesitant to do it if that wasn't a helpful step towards deciding this feature was actually desirable.

**Comment by Gabriella439** [comment]
@mattpolzin: In my experience, usually if someone is willing to put in the effort to standardize something that's a pretty good sign that the feature would be widely well-received and somewhat maintainable for the various implementations.  Usually if something is hard for implementations to implement it's also typically hard to standardize, so that's the natural check on the process.
